stdlib/socket/src/main/java/org/ballerinalang/stdlib/socket/tcp/SelectorManager.java

```diff
@@ -0,0 +1,238 @@
+                if (select == 0) {
+                    continue;
+                }
+                Iterator<SelectionKey> keyIterator = selector.selectedKeys().iterator();
+                while (keyIterator.hasNext()) {
+                    SelectionKey key = keyIterator.next();
+                    keyIterator.remove();
+                    performAction(key);
+                }
+            } catch (Throwable e) {
+                log.error("An error occurred in selector loop: " + e.getMessage(), e);
+            }
+        }
+    }
+
+    /*
+    Channel registration has to be done in the same thread that selector loops runs.
+     */
+    private void registerChannels() {
+        ChannelRegisterCallback channelRegisterCallback;
+        while ((channelRegisterCallback = registerPendingSockets.poll()) != null) {
+            try {
+                SocketService socketService = channelRegisterCallback.getSocketService();
+                socketService.getSocketChannel()
+                        .register(selector, channelRegisterCallback.getInitialInterest(), socketService);
+            } catch (ClosedChannelException e) {
+                channelRegisterCallback.notifyFailure("Socket already closed");
+            }
+            channelRegisterCallback
+                    .notifyRegister(channelRegisterCallback.getInitialInterest() == SelectionKey.OP_READ);
+        }
+    }
+
+    private void performAction(SelectionKey key) {
+        if (!key.isValid()) {
+            key.cancel();
+        } else if (key.isAcceptable()) {
+            onAccept(key);
+        } else if (key.isReadable()) {
+            onReadReady(key);
+        }
+    }
+
+    private void onAccept(SelectionKey key) {
+        SocketService socketService = (SocketService) key.attachment();
+        try {
+            ServerSocketChannel server = (ServerSocketChannel) socketService.getSocketChannel();
+            SocketChannel client = server.accept();
+            client.configureBlocking(false);
+            // Registering the channel against the selector directly without going through the queue,
+            // since we are in same thread.
+            client.register(selector, OP_READ, new SocketService(client, socketService.getResources()));
+            SelectorDispatcher.invokeOnAccept(socketService, client);
+        } catch (ClosedByInterruptException e) {
+            SelectorDispatcher.invokeOnError(socketService, "Client accept interrupt by another process");
+        } catch (AsynchronousCloseException e) {
+            SelectorDispatcher.invokeOnError(socketService, "Client closed by another process");
+        } catch (ClosedChannelException e) {
+            SelectorDispatcher.invokeOnError(socketService, "Client is already closed");
+        } catch (IOException e) {
+            log.error("An error occurred while accepting new client", e);
+            SelectorDispatcher.invokeOnError(socketService, "Unable to accept a new client");
```

Comment:
Why are we logging in the error only in this case?

Reply:
We are not in the position to pass the error message directly to the ballerina user. But for the troubleshoot purposes, we need this to log.